Realm/Realm/Handles/SessionHandle.cs

```diff
@@ -224,14 +261,86 @@ private static void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorC
                     exception = new SessionException(messageString, errorCode);
                 }
 
-                Session.RaiseError(session, exception);
+                if (syncConfigurationBase.OnSessionError != null)
+                {
+                    syncConfigurationBase.OnSessionError.Invoke(session, exception);
+                }
+                else
+                {
+                    // after deprecation: this will need to go when Session.Error is fully deprecated
+                    Session.RaiseError(session, exception);
+                }
             }
             catch (Exception ex)
             {
                 Logger.Default.Log(LogLevel.Warn, $"An error has occurred while handling a session error: {ex}");
             }
         }
 
+        [MonoPInvokeCallback(typeof(NativeMethods.NotifyBeforeClientReset))]
+        private static bool NotifyBeforeClientReset(IntPtr beforeFrozen, IntPtr managedSyncConfigurationHandle)
+        {
+            var syncConfigHandle = GCHandle.FromIntPtr(managedSyncConfigurationHandle);
+            var syncConfiguration = (SyncConfigurationBase)syncConfigHandle.Target;
+
+            if (syncConfiguration.ClientResetHandler == null)
+            {
+                return true;
+            }
+
+            var discardLocalResetHandler = (DiscardLocalResetHandler)syncConfiguration.ClientResetHandler;
+            var schema = syncConfiguration.GetSchema();
+
+            try
+            {
+                if (discardLocalResetHandler.OnBeforeReset != null)
+                {
+                    using var realmBefore = new Realm(new UnownedRealmHandle(beforeFrozen), syncConfiguration, schema);
```

Comment:
The only built `RealmMetadata` that I could find is in `Realm`. Otherwise metadata is built inside the constructor of `Realm`. I don't know a more efficient way of building `RealmMetadata` then what the constructor already does. So I need a little more about what you suggest

Reply:
Oh right... we don't store the metadata in the configuration. Meh... it's fine then, let's keep it as is.